// @orca-so/whirlpools-sdk/dist/types/public/anchor-types.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toTick = exports.WHIRLPOOL_ACCOUNT_SIZE = exports.WHIRLPOOL_CODER = exports.WHIRLPOOL_IDL = exports.AccountName = void 0;
exports.getAccountSize = getAccountSize;
const anchor_1 = require("@coral-xyz/anchor");
const anchor_2 = require("@coral-xyz/anchor");
const idl_1 = require("@coral-xyz/anchor/dist/cjs/idl");
const whirlpool_json_1 = __importDefault(require("../../artifacts/whirlpool.json"));
var AccountName;
(function (AccountName) {
    AccountName["WhirlpoolsConfig"] = "whirlpoolsConfig";
    AccountName["Position"] = "position";
    AccountName["TickArray"] = "tickArray";
    AccountName["DynamicTickArray"] = "dynamicTickArray";
    AccountName["Whirlpool"] = "whirlpool";
    AccountName["FeeTier"] = "feeTier";
    AccountName["PositionBundle"] = "positionBundle";
    AccountName["WhirlpoolsConfigExtension"] = "whirlpoolsConfigExtension";
    AccountName["TokenBadge"] = "tokenBadge";
    AccountName["LockConfig"] = "lockConfig";
    AccountName["Oracle"] = "oracle";
    AccountName["AdaptiveFeeTier"] = "adaptiveFeeTier";
})(AccountName || (exports.AccountName = AccountName = {}));
exports.WHIRLPOOL_IDL = whirlpool_json_1.default;
exports.WHIRLPOOL_CODER = new anchor_2.BorshAccountsCoder((0, idl_1.convertIdlToCamelCase)(exports.WHIRLPOOL_IDL));
function getAccountSize(accountName) {
    const size = exports.WHIRLPOOL_CODER.size(accountName);
    return size + RESERVED_BYTES[accountName];
}
const RESERVED_BYTES = {
    [AccountName.WhirlpoolsConfig]: 0,
    [AccountName.Position]: 0,
    [AccountName.TickArray]: 0,
    [AccountName.DynamicTickArray]: 0,
    [AccountName.Whirlpool]: 0,
    [AccountName.FeeTier]: 0,
    [AccountName.PositionBundle]: 64,
    [AccountName.WhirlpoolsConfigExtension]: 512,
    [AccountName.TokenBadge]: 128,
    [AccountName.LockConfig]: 128,
    [AccountName.Oracle]: 0,
    [AccountName.AdaptiveFeeTier]: 128,
};
exports.WHIRLPOOL_ACCOUNT_SIZE = getAccountSize(AccountName.Whirlpool);
const toTick = (tick) => {
    if ("uninitialized" in tick) {
        return {
            initialized: false,
            liquidityNet: new anchor_1.BN(0),
            liquidityGross: new anchor_1.BN(0),
            feeGrowthOutsideA: new anchor_1.BN(0),
            feeGrowthOutsideB: new anchor_1.BN(0),
            rewardGrowthsOutside: [new anchor_1.BN(0), new anchor_1.BN(0), new anchor_1.BN(0)],
        };
    }
    return {
        initialized: true,
        ...tick.initialized[0],
    };
};
exports.toTick = toTick;
